Library/Homebrew/formula_auditor.rb

```diff
@@ -848,5 +855,16 @@ def new_formula_problem(message, location: nil)
     def head_only?(formula)
       formula.head && formula.stable.nil?
     end
+
+    def linux_only_gcc_dep?(formula)
+      # TODO: Make this check work when running on Linux and not simulating macOS too.
+      return false unless Homebrew::SimulateSystem.simulating_or_running_on_macos?
+
+      formula_hash = formula.to_hash_with_variations
+      deps = formula_hash["dependencies"]
+      linux_deps = formula_hash.dig("variations", :x86_64_linux, "dependencies")
+
+      deps.exclude?("gcc") && linux_deps&.include?("gcc")
```

Comment:
I wonder if something like this would allow the check to work on both platforms?

```suggestion
      formula_hash = formula.to_hash_with_variations
      linux_hash = formula_hash.merge formula_hash.dig("variations", :x86_64_linux)
      macos_hash = formula_hash.merge formula_hash.dig("variations", :monterey)

      linux_hash["dependencies"].exclude?("gcc") && macos_hash["dependencies"].include?("gcc")
```

Of course, the edge case here is that if `gcc` is a dependency on Linux and some (but not all) macOS versions, there's a risk for a false positive.

Reply:
False positives are pretty bad (they will prevent any testing from being done when `tap_syntax` errors), so I'd like to avoid them.

I had an implementation in mind where I check all keys in `formula_hash["variations"]` and throw out any `:x86_64_linux` whenever we're on Linux (or simulating it), and then check if `gcc` appears anywhere there.

But it seemed a bit complicated and I'm not super convinced we need it yet...

Reply:
> False positives are pretty bad (they will prevent any testing from being done when `tap_syntax` errors), so I'd like to avoid them, even if it means this test won't run properly on Linux.

To be clear, I don't think that suggestion includes any false positives that the current solution doesn't include. Say I'm running (or simulating) `arm64_monterey` and the `gcc` dep is defined using `on_system :linux, macos: :catalina_or_older`.

As the PR currently stands, a problem would be reported since the dep is present on Linux and not present on the current simulated or real version. However, it's not really a Linux-only dep since it's also required on `catalina_or_older`

Reply:
Good point. I'll think about how to deal with that, but I don't think it'll come up in Homebrew/core, so this doesn't need to block this for now. (Or, if I'm wrong: we can revert this.)